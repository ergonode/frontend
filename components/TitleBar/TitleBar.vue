/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <div class="title-bar">
        <TitleBarDetails
            :title="title"
            :icon="icon"
            :is-breadcrumb="breadcrumbs.length !== 0"
            @navigateback="onClick">
            <template #breadcrumb>
                <TitleBarBreadcrumb
                    v-for="(breadcrumb, index) in breadcrumbs"
                    :key="index"
                    :breadcrumb="breadcrumb" />
            </template>
            <template #badge>
                <InformationBadge
                    v-if="status"
                    :background="status.color"
                    :color="statusTextContrastRation"
                    :title="status.code" />
                <InformationBadge
                    v-if="isReadOnly"
                    :background="blueColor"
                    :color="whiteColor"
                    title="READ ONLY">
                    <template #prepend>
                        <IconLock
                            class="badge__icon"
                            size="24"
                            :fill-color="whiteColor" />
                    </template>
                </InformationBadge>
            </template>
        </TitleBarDetails>
        <TitleBarActions :buttons="buttons" />
    </div>
</template>

<script>
import { WHITE, BLUE, GRAPHITE_DARK } from '~/assets/scss/_variables/_colors.scss';
import {
    hexToRGB, calculateRelativeLuminance, calculateContrastRatio,
} from '~/model/color/ColorContrast';

export default {
    name: 'TitleBar',
    components: {
        TitleBarDetails: () => import('~/components/TitleBar/TitleBarDetails'),
        TitleBarActions: () => import('~/components/TitleBar/TitleBarActions'),
        TitleBarBreadcrumb: () => import('~/components/TitleBar/TitleBarBreadcrumb'),
        InformationBadge: () => import('~/components/Badges/InformationBadge'),
        IconLock: () => import('~/components/Icon/Feedback/IconLock'),
    },
    props: {
        title: {
            type: String,
            required: true,
        },
        buttons: {
            type: Array,
            default: () => [],
        },
        status: {
            type: Object,
            default: () => {},
        },
        icon: {
            type: String,
            default: null,
        },
        breadcrumbs: {
            type: Array,
            default: () => [],
        },
        isReadOnly: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        whiteColor() {
            return WHITE;
        },
        blueColor() {
            return BLUE;
        },
        statusTextContrastRation() {
            const convertedRgbText = hexToRGB(WHITE);
            const convertedRgbBg = hexToRGB(this.status.color);
            const relativeLuminance = calculateRelativeLuminance(convertedRgbText);
            const relativeLuminanceBackground = calculateRelativeLuminance(convertedRgbBg);
            const contrastRatio = calculateContrastRatio(
                relativeLuminance,
                relativeLuminanceBackground,
            );

            return contrastRatio > 4.5 ? WHITE : GRAPHITE_DARK;
        },
    },
    methods: {
        onClick() {
            this.$emit('navigateback');
        },
    },
};
</script>

<style lang="scss" scoped>
    .title-bar {
        display: flex;
        flex: 0 0 auto;
        justify-content: space-between;
        align-items: center;
        min-height: 32px;
        padding: 24px;

        .badge__icon {
            margin: -6px 0 -6px -8px;
        }
    }

</style>
