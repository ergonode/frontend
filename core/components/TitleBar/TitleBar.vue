/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <div class="title-bar">
        <TitleBarHeader :title="title">
            <template #prepend>
                <FabButton
                    v-if="isNavigationBack"
                    @click.native="onClick">
                    <template #icon="{ color }">
                        <IconArrowPointer :fill-color="color" />
                    </template>
                </FabButton>
            </template>
            <template #append>
                <slot name="prependBadge" />
                <InformationIconBadge
                    v-if="isReadOnly"
                    :background="blueColor"
                    :color="whiteColor"
                    title="READ ONLY">
                    <template #icon>
                        <IconLock
                            size="24"
                            :fill-color="whiteColor" />
                    </template>
                </InformationIconBadge>
            </template>
        </TitleBarHeader>
        <TitleBarActions>
            <slot name="subActions" />
            <slot name="mainAction" />
        </TitleBarActions>
    </div>
</template>

<script>
import { WHITE, BLUE, GREEN } from '~/assets/scss/_variables/_colors.scss';

export default {
    name: 'TitleBar',
    components: {
        TitleBarHeader: () => import('~/core/components/TitleBar/TitleBarHeader'),
        TitleBarActions: () => import('~/core/components/TitleBar/TitleBarActions'),
        InformationIconBadge: () => import('~/core/components/Badges/InformationIconBadge'),
        IconLock: () => import('~/components/Icon/Feedback/IconLock'),
        IconArrowPointer: () => import('~/components/Icon/Arrows/IconArrowPointer'),
        FabButton: () => import('~/core/components/Buttons/FabButton'),
    },
    props: {
        title: {
            type: String,
            required: true,
        },
        icon: {
            type: String,
            default: null,
        },
        isNavigationBack: {
            type: Boolean,
            default: false,
        },
        isReadOnly: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        whiteColor() {
            return WHITE;
        },
        blueColor() {
            return BLUE;
        },
        greenColor() {
            return GREEN;
        },
    },
    methods: {
        onClick() {
            this.$emit('navigateBack');
        },
    },
};
</script>

<style lang="scss" scoped>
    .title-bar {
        display: flex;
        flex: 0 0 auto;
        justify-content: space-between;
        align-items: center;
        min-height: 32px;
        padding: 24px;
    }
</style>
