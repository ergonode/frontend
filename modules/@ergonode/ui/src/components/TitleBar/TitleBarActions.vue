/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <div class="title-bar-actions">
        <slot />
    </div>
</template>

<script>
export default {
    name: 'TitleBarActions',
};
</script>

<style lang="scss" scoped>
    .title-bar-actions {
        display: grid;
        grid-auto-flow: column;
        grid-column-gap: 16px;
        align-items: center;
    }
</style>
