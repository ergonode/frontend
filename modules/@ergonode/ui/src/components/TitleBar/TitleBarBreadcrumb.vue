/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <NuxtLink
        class="breadcrumb"
        :to="breadcrumb.route">
        <Component
            :is="breadcrumbIconComponent"
            :fill-color="greenColor"
            width="16"
            height="16" />
        <span
            class="breadcrumb__title"
            v-text="breadcrumb.title" />
    </NuxtLink>
</template>

<script>
import {
    GREEN,
} from '@UI/assets/scss/_js-variables/colors.scss';

export default {
    name: 'TitleBarBreadcrumb',
    props: {
        /**
         * Breadcrumb data model
         */
        breadcrumb: {
            type: Object,
            required: true,
        },
    },
    computed: {
        greenColor() {
            return GREEN;
        },
        breadcrumbIconComponent() {
            return () => import(`@Core/components/Icons/Menu/Icon${this.breadcrumb.icon}`);
        },
    },
};
</script>

<style lang="scss" scoped>
    .breadcrumb {
        display: flex;
        align-items: center;
        grid-column: 2 / 3;
        grid-row: 2 / 3;
        margin-left: 8px;
        text-decoration: none;

        &:active {
            color: inherit;
        }

        &__title {
            margin-left: 4px;
            color: $GRAPHITE_DARK;
            font: $FONT_MEDIUM_14_20;
        }

        &:hover {
            text-decoration: underline;
        }
    }
</style>
