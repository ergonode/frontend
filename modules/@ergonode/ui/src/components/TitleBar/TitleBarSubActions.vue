/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <div class="title-bar-sub-actions">
        <slot />
    </div>
</template>

<script>
export default {
    name: 'TitleBarSubActions',
};
</script>

<style lang="scss" scoped>
    .title-bar-sub-actions {
        display: grid;
        grid-auto-flow: column;
        grid-column-gap: 8px;
        align-items: center;
    }
</style>
