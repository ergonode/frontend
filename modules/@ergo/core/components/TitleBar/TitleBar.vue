/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <div class="title-bar">
        <TitleBarHeader :title="title">
            <template #prepend>
                <Fab
                    v-if="isNavigationBack"
                    @click.native="onClick">
                    <template #icon="{ color }">
                        <IconArrowPointer :fill-color="color" />
                    </template>
                </Fab>
            </template>
            <template #append>
                <slot name="prependBadge" />
                <InformationIconBadge
                    v-if="isReadOnly"
                    :background="blueColor"
                    :color="whiteColor"
                    title="READ ONLY">
                    <template #icon="{ color }">
                        <IconLock
                            size="24"
                            :fill-color="color" />
                    </template>
                </InformationIconBadge>
            </template>
        </TitleBarHeader>
        <TitleBarActions>
            <slot name="subActions" />
            <slot name="mainAction" />
        </TitleBarActions>
    </div>
</template>

<script>
import { WHITE, BLUE, GREEN } from '@Core/assets/scss/_js-variables/colors.scss';
import TitleBarHeader from '@Core/components/TitleBar/TitleBarHeader';
import TitleBarActions from '@Core/components/TitleBar/TitleBarActions';
import Fab from '@Core/components/Buttons/Fab';

export default {
    name: 'TitleBar',
    components: {
        TitleBarHeader,
        TitleBarActions,
        Fab,
        IconLock: () => import('@Core/components/Icons/Feedback/IconLock'),
        IconArrowPointer: () => import('@Core/components/Icons/Arrows/IconArrowPointer'),
        InformationIconBadge: () => import('@Core/components/Badges/InformationIconBadge'),
    },
    props: {
        title: {
            type: String,
            required: true,
        },
        icon: {
            type: String,
            default: null,
        },
        isNavigationBack: {
            type: Boolean,
            default: false,
        },
        isReadOnly: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        whiteColor() {
            return WHITE;
        },
        blueColor() {
            return BLUE;
        },
        greenColor() {
            return GREEN;
        },
    },
    methods: {
        onClick() {
            this.$emit('navigateBack');
        },
    },
};
</script>

<style lang="scss" scoped>
    .title-bar {
        display: flex;
        flex: 0 0 auto;
        justify-content: space-between;
        align-items: center;
        min-height: 32px;
        padding: 24px;
    }
</style>
