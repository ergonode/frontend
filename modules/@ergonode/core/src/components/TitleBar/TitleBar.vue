/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <header class="title-bar">
        <TitleBarHeader :title="title">
            <template #prepend>
                <slot name="prependHeader" />
            </template>
            <template #append>
                <slot name="prependBadge" />
                <InformationIconBadge
                    v-if="isReadOnly"
                    :background="blueColor"
                    :color="whiteColor"
                    title="READ ONLY">
                    <template #icon="{ color }">
                        <IconLock :fill-color="color" />
                    </template>
                </InformationIconBadge>
            </template>
        </TitleBarHeader>
        <TitleBarActions>
            <slot name="subActions" />
            <slot name="mainAction" />
        </TitleBarActions>
    </header>
</template>

<script>
import {
    BLUE,
    GREEN,
    WHITE,
} from '@Core/assets/scss/_js-variables/colors.scss';
import InformationIconBadge from '@Core/components/Badges/InformationIconBadge';
import IconLock from '@Core/components/Icons/Feedback/IconLock';
import TitleBarActions from '@Core/components/TitleBar/TitleBarActions';
import TitleBarHeader from '@Core/components/TitleBar/TitleBarHeader';

export default {
    name: 'TitleBar',
    components: {
        TitleBarHeader,
        TitleBarActions,
        IconLock,
        InformationIconBadge,
    },
    props: {
        /**
         * The title of the component
         */
        title: {
            type: String,
            required: true,
        },
        /**
         * Determines if the context is read only
         */
        isReadOnly: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        whiteColor() {
            return WHITE;
        },
        blueColor() {
            return BLUE;
        },
        greenColor() {
            return GREEN;
        },
    },
};
</script>

<style lang="scss" scoped>
    .title-bar {
        display: flex;
        flex: 0 0 auto;
        justify-content: space-between;
        align-items: center;
        min-height: 32px;
        padding: 24px;
    }
</style>
