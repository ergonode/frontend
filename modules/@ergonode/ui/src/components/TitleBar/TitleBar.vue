/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <header class="title-bar">
        <TitleBarHeader :title="title">
            <template #prepend>
                <slot name="prependHeader" />
            </template>
            <template #append>
                <slot name="prependBadge" />
                <InformationIconBadge
                    v-if="isReadOnly"
                    :background="blueColor"
                    :color="whiteColor"
                    title="READ ONLY">
                    <template #icon="{ color }">
                        <IconLock :fill-color="color" />
                    </template>
                </InformationIconBadge>
            </template>
        </TitleBarHeader>
        <TitleBarActions>
            <slot name="subActions" />
            <slot name="mainAction" />
        </TitleBarActions>
    </header>
</template>

<script>
import {
    BLUE,
    GREEN,
    WHITE,
} from '@UI/assets/scss/_js-variables/colors.scss';
import InformationIconBadge from '@UI/components/Badges/InformationIconBadge';
import IconLock from '@UI/components/Icons/Feedback/IconLock';
import TitleBarActions from '@UI/components/TitleBar/TitleBarActions';
import TitleBarHeader from '@UI/components/TitleBar/TitleBarHeader';

export default {
    name: 'TitleBar',
    components: {
        TitleBarHeader,
        TitleBarActions,
        IconLock,
        InformationIconBadge,
    },
    props: {
        /**
         * The title of the component
         */
        title: {
            type: String,
            required: true,
        },
        /**
         * Determines if the context is read only
         */
        isReadOnly: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        whiteColor() {
            return WHITE;
        },
        blueColor() {
            return BLUE;
        },
        greenColor() {
            return GREEN;
        },
    },
};
</script>

<style lang="scss" scoped>
    .title-bar {
        display: flex;
        flex: 0 0 auto;
        justify-content: space-between;
        align-items: center;
        min-height: 32px;
        padding: 24px;
    }
</style>
