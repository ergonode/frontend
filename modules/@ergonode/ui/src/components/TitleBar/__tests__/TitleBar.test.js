/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
import TitleBar from '@UI/components/TitleBar/TitleBar';
import {
    shallowMount,
} from '@vue/test-utils';

describe('TitleBar/TitleBar', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallowMount(TitleBar, {
            propsData: {
                title: 'test',
                isReadOnly: true,
            },
        });
    });

    it('Component is named well', () => {
        expect(typeof TitleBar.name).toBe('string');
        expect(TitleBar.name).toEqual('TitleBar');
    });

    it('TitleBar is rendered with information badge', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
