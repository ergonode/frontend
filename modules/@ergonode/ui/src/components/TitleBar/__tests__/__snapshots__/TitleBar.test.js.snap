// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleBar/TitleBar TitleBar is rendered with information badge 1`] = `
<header data-cy="title-bar" class="title-bar">
  <titlebarheader-stub title="test"></titlebarheader-stub>
  <titlebaractions-stub> </titlebaractions-stub>
</header>
`;
